### ISSUE TITLE
running in parallel on same OS & browser

### ISSUE THREAD

**Issue by hamiltonf** [theintern/intern#391]
Hi,

As we are working on CI/CD and we have more the 500 functional tests on top of a 5000 unit, we need the functional tests to run in parallel on around 100 vm's that have the same configuration of OS and Browsers. didn't find a way to do it with intern, so if it's possible to add it?

Thanks.

**Comment by foooocus** [comment]
Looking for the same answer, is there a way to run parallel tests on same platform?

**Comment by sholladay** [comment]
A hackaround is to launch two instances of Intern with two different configuration files, set up to divide your suites between them.

Not saying it's fun. But, do-able.

**Comment by DaveStein** [comment]
@sholladay I am confused about the default port of 9000. If I try to run two instances, the second instant blows up cause 9000 is in use. And in my very noob-like guessing, I changed proxy and proxyUrl to use 9001 instead of 9000 but then that finds nothing.

**Comment by sholladay** [comment]
Yeah, they will need to be on different ports, for sure. I suggest you put what you tried in a [gist](https://gist.github.com/) and post a link to it, maybe I can spot what is wrong with it. Even better would be to post about it on [StackOverflow](http://stackoverflow.com/) so it gets more visibility.

**Comment by DaveStein** [comment]
Ha in writing to SO I realize I made a typo! One extra zero.

```
#!/bin/bash
#!/usr/bin/env node

./node_modules/.bin/intern-runner config=tests/intern \
functionalSuites=tests/functional/sample1 \
proxyPort=9001 \
proxyUrl=http://localhost:9001/ &

./node_modules/.bin/intern-runner config=tests/intern \
functionalSuites=tests/functional/sample2 \
proxyPort=9002 \
proxyUrl=http://localhost:9002/ &
```

That hack worked. Mind you I never write bash so I gotta go look up how to make an array of suites to just iterate rather than copy paste over and over ;)

**Comment by DaveStein** [comment]
```
#!/bin/bash
#!/usr/bin/env node

declare -a suites=( \
  "example1" \
  "example2" \
)

arraylength=${#suites[@]};

mkdir ./tests/functional/screenshots/$datetime/;

for (( i=0; i<${arraylength}; i++ ));
do
  p=$(expr $i + 1);
  ./node_modules/.bin/intern-runner config=tests/intern \
  functionalSuites=tests/functional/${suites[$i]} \
  proxyPort=900$p \
  proxyUrl=http://localhost:900$p/ &
done;

```

**Comment by sholladay** [comment]
That's a solid recipe. Maybe do a PR to [intern-examples](https://github.com/theintern/intern-examples) with it.

FYI, you should replace the first two lines with `#!/usr/bin/env bash`

**Comment by DaveStein** [comment]
What does that change exactly? As you can tell I did a bunch of copy pasting from the internet :)

If I made a PR, do you think that's just bash-example? lol not sure where i'd put it cause I'm not using anything whatsoever except intern and this script - loading up like google in my test for example

**Comment by sholladay** [comment]
That is a unix [shebang](https://en.wikipedia.org/wiki/Shebang_(Unix)), only the first one has any effect. The second one you had is super useful for writing command line programs with Node, but your script is meant to be run with bash. And again, it only makes sense to have [one at a time](http://www.tldp.org/LDP/abs/html/sha-bang.html). The reason for avoiding `#!/bin/bash` and using `#!/usr/bin/env bash` instead is because the first one "hard codes" the location of bash itself in there, while the second one effectively asks the [$PATH](https://en.wikipedia.org/wiki/PATH_(variable)) to find it. This means that it will run with the user's preferred copy of bash.

Most computers do have a bash at `/bin/bash`, but on OS X it is ancient, and rather than overwriting Apple's copy, people use something like [Homebrew](https://github.com/Homebrew/homebrew) to install a modern version. And you generally want to respect that. As with most things, there are [pros and cons](http://unix.stackexchange.com/questions/29608/why-is-it-better-to-use-usr-bin-env-name-instead-of-path-to-name-as-my), but I recommend it.

As for the PR, I am sure @vladikoff will help you figure out the best way to showcase this. But `bash-example` sounds good to me. :)

**Comment by DaveStein** [comment]
Expect PR in a couple days or so :) @sholladay apparently using selenium with java, you can even run tests from same suite in parallel. this gets me halfway there

**Comment by szydan** [comment]
@DaveStein Is is necessary to set both 
**proxyPort** and **proxyUrl** or it is enough to set only **proxyPort**

**Comment by jason0x43** [comment]
@szydan `proxyPort` and `proxyUrl` are separate entities. Intern always serves on localhost, and `proxyPort` is the port on localhost that Intern should use. `proxyUrl`, on the other hand, is the URL that a remote client (like `client.html`) should use to communicate with Intern. This may or may not be `localhost:<proxyPort>` (for example, it could be different if Intern were running behind a proxy from the client's point of view).

**Comment by jason0x43** [comment]
Closing this in favor of #779
